misc/quick-protobuf-codec/src/lib.rs

```diff
@@ -53,24 +95,238 @@ where
+            Err(unsigned_varint::decode::Error::Insufficient) => return Ok(None),
+            Err(e) => return Err(Error(io::Error::new(io::ErrorKind::InvalidData, e))),
         };
 
-        let mut reader = BytesReader::from_bytes(&msg);
-        let message = Self::Item::from_reader(&mut reader, &msg)
-            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
+        if message_length > self.max_message_len_bytes {
+            return Err(Error(io::Error::new(
+                io::ErrorKind::PermissionDenied,
+                format!(
+                    "message with {message_length}b exceeds maximum of {}b",
+                    self.max_message_len_bytes
+                ),
+            )));
+        }
+
+        // Compute how many bytes the varint itself consumed.
+        let varint_length = src.len() - remaining.len();
+
+        // Ensure we can read an entire message.
+        if src.len() < (message_length + varint_length) {
+            return Ok(None);
+        }
+
+        // Safe to advance buffer now.
+        src.advance(varint_length);
+
+        let message = src.split_to(message_length);
+
+        let mut reader = BytesReader::from_bytes(&message);
+        let message = Self::Item::from_reader(&mut reader, &message)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
+
         Ok(Some(message))
     }
 }
 
+struct MaybeUninitWriterBackend<'a> {
+    memory: &'a mut [MaybeUninit<u8>],
+    written: &'a mut usize,
+}
+
+impl<'a> MaybeUninitWriterBackend<'a> {
+    fn new(memory: &'a mut [MaybeUninit<u8>], written: &'a mut usize) -> Self {
+        Self { memory, written }
+    }
+}
+
+impl<'a> WriterBackend for MaybeUninitWriterBackend<'a> {
+    fn pb_write_u8(&mut self, x: u8) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&[x])
+    }
+
+    fn pb_write_u32(&mut self, x: u32) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&x.to_le_bytes())
+    }
+
+    fn pb_write_i32(&mut self, x: i32) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&x.to_le_bytes())
+    }
+
+    fn pb_write_f32(&mut self, x: f32) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&x.to_le_bytes())
+    }
+
+    fn pb_write_u64(&mut self, x: u64) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&x.to_le_bytes())
+    }
+
+    fn pb_write_i64(&mut self, x: i64) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&x.to_le_bytes())
+    }
+
+    fn pb_write_f64(&mut self, x: f64) -> quick_protobuf::Result<()> {
+        self.pb_write_all(&x.to_le_bytes())
+    }
+
+    fn pb_write_all(&mut self, buf: &[u8]) -> quick_protobuf::Result<()> {
+        if self.memory.len() - *self.written < buf.len() {
+            return Err(quick_protobuf::errors::Error::UnexpectedEndOfBuffer);
+        }
+
+        for b in buf {
+            self.memory[*self.written].write(*b);
+            *self.written += 1;
+        }
```

Comment:
Unfortunately, `MaybeUninit::write_slice` is still unstable but we could inline it here? It might be more performant for large buffers.